tempesta_fw/sock.c

```diff
@@ -440,6 +440,13 @@ ss_do_close(struct sock *sk)
 	}
 }
 
+static void
+ss_do_close(struct sock *sk)
+{
+	__ss_do_close(sk);
+	sock_put(sk);
+}
+
```

Comment:
`ss_close()` is called where is can be used in process context, so in includes additional locking. `ss_do_close()` is called in SoftIRQ context, and it does not need any extra locking around it, the socket is most probably locked already anyway. `__ss_do_close()` is the main body of the close function, and it is therefore called from both `ss_close()` and `ss_do_close()`.

`sock_put()` should be called outside of locks, that's why it's not in `__ss_do_close()` (or `ss_do_close()` before this patch).